File: LeetCode/200. Number of Islands/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        int islands = 0;

        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(grid[i][j] == '1'){
                    sinkIslands(grid, i,j);
                    islands++;
                }
            }
        }

        return islands;
    }

    void sinkIslands(vector<vector<char>>& grid, int i, int j){

        grid[i][j] = '0';

        if(i-1 >= 0 && grid[i-1][j] == '1') 
            sinkIslands(grid, i-1, j);
        
        if(i+1 < grid.size() && grid[i+1][j] == '1') 
            sinkIslands(grid, i+1, j);
        
        if(j-1 >= 0 && grid[i][j-1] == '1')
            sinkIslands(grid, i, j-1);

        if(j+1 < grid[0].size() && grid[i][j+1] == '1')
            sinkIslands(grid, i, j+1);
    }
};

File: LeetCode/200. Number of Islands/solution.py

'''
Runtime
243
ms
Beats
53.68%
of users with Python3
Memory
18.78
MB
Beats
97.75%
of users with Python3
'''
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        n = len(grid)
        m = len(grid[0])
        islands = 0

        for i in range(n):
            for j in range(m):
                if grid[i][j] == '1':
                    self.sinkIsland(grid, i, j)
                    islands += 1
        
        return islands
    

    def sinkIsland(self, grid, i, j):

        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]):
            return
        
        if grid[i][j] == '0':
            return
        
        grid[i][j] = '0'

        dx = [-1, 0, 1, 0]
        dy = [0, -1, 0, 1]
        for d in range(4):
            self.sinkIsland(grid, i+dx[d], j+dy[d])

        